state/execution.go

```diff
@@ -326,29 +326,29 @@ func updateValidators(currentSet *types.ValidatorSet, abciUpdates []abci.Validat
 	if err != nil {
 		return err
 	}
-
-	// these are tendermint types now
+	totalVotingPower := currentSet.TotalVotingPower()
 	for _, valUpdate := range updates {
 		if valUpdate.VotingPower < 0 {
 			return fmt.Errorf("Voting power can't be negative %v", valUpdate)
 		}
 
 		address := valUpdate.Address
 		_, val := currentSet.GetByAddress(address)
-		if valUpdate.VotingPower == 0 {
-			// remove val
+		if valUpdate.VotingPower == 0 { // remove val
 			_, removed := currentSet.Remove(address)
 			if !removed {
 				return fmt.Errorf("Failed to remove validator %X", address)
 			}
-		} else if val == nil {
-			// add val
+		} else if val == nil { // add val
+			// TODO: issue #1558 update spec according to the following:
+			// Set Accum to -totalVotingPower to make sure validators can't unbond/rebond to reset their (potentially
+			// previously negative) Accum to zero.
+			valUpdate.Accum = -totalVotingPower
```

Comment:
I still think we should multiply by a constant (maybe 2 or 3)

Without a constant, it will be advantageous to still unbond in a lot of scenarios.

Reply:
The scenario is what you describe in https://github.com/tendermint/tendermint/issues/2718#issuecomment-437187666 right?

By using a multiplicative factor, e.g. 2, this would rather mean you can split up in 3 or more validators and do a similar thing you described, is that right? I can change it to 3. And we can do a proper decoupling after launch, OK? @ebuchman @ValarDragon @milosevic @cwgoes

Reply:
No, the attack is mitigated in all cases by using such a constant. (It is not eliminated, solely mitigated) The mitigation comes from unbonding always having a cost for a validator that would normally be proposing, whereas in the described system, there is no cost / it may be profitable to do so anyway. (As suppose your the lowest, then unbonding / rebonding still helps you become not the lowest)

The idea being simulated is when you initially bond, there is a period where you aren't eligible to propose and your accum isn't increasing, but you do contribute to total stake. Perhaps its better to just do that directly.

Reply:
I guess. I feel like its worked out in the issue, but given the discussion there perhaps I didn't explain it well enough. Also correctness can't really be proven, since the algorithm itself is quite biasable. The constant was just meant as an easy way of simulating you not getting to be a proposer for the first n blocks after you bond, so that would still be better.

Reply:
Thoughts on https://github.com/tendermint/tendermint/issues/2718#issuecomment-438736669 @ValarDragon?

Reply:
I think we could add a constant here, although I think `2` or `3` is too high - `1.1` or `1.2` should be sufficient to mitigate the discussed attack (which I am not yet convinced is profitable even with `c = 1`) - unless there's some part of it I don't understand that would require a high `c` to mitigate (cc @ValarDragon)?

Reply:
You can't multiply by 1.1 or 1.2 without casting to float, and not all CPU's implement IEEE 754. That means we'd have to use an in-software IEEE 754 float. Its do-able, its just always annoying to use them. Or maybe we could use Unum's :D

Reply:
Hmm, how about multiply by 10 and divide by 9?

Reply:
Hrmm, we can multiply by 1.125 quite efficiently, `x + (x >> 3)`

Reply:
that works!

Reply:
OK, `x + (x >> 3)` works but we'd need to make sure we do not overflow here. `vals.TotalVotingPower()` is only bound by `MaxInt64` (via clipping). Computing `x + (x >> 3)` could potentially overflow. A consistent way to deal with this would be to clip (again) in the case of an overflow. Alternatively, we could enforce a stricter bound on totalVotingPower.

Reply:
[]()